ixmp/core.py

```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
         """
         self._jobj.addUnitToDB(unit, comment)
 
+    def regions(self):
+        """Return all regions defined for the IAMC-style timeseries format
+        including known synonyms.
+
+        Returns
+        -------
+        pd.Dataframe
+        """
+        lst = []
+        for r in self._jobj.listNodes('%'):
+            n, p, h = (r.getName(), r.getParent(), r.getHierarchy())
+            lst.extend([(n, None, p, h)])
+            lst.extend([(s, n, p, h) for s in (r.getSynonyms() or [])])
+        region = pd.DataFrame(lst)
+        region.columns = ['region', 'mapped_to', 'parent', 'hierarchy']
+        return region
+
+    def add_region(self, region, hierarchy, parent='World'):
+        """Define a region including a hierarchy level and a 'parent' region.
+        *Before adding a region, please use `regions()` and check whether the
+        region already exists with a different spelling.
+        If so, use `add_region_synonym()` instead.
+
+        Parameters
+        ----------
+        region : str
+            Name of the region.
+        hierarchy : str
+            Hierarchy level of the region (e.g., country, R11, basin)
```

Comment:
Does the hierarchy level need to exist? Maybe say (I'm guessing) "If the given level does not exist, it is created."

A question as an aside: how is the hierarchy of levels set? E.g. where/how is it specified that an 'R11' contains 1 or more 'country'; or that a particular country is in a particular R11?

Reply:
The hierarchy level and the mapping to regions is currently purely informative and there are no validation steps - the database will accept any values and any mapping (as long as the `parent` region exists in the `region` column).